src/vis/density/HexagonalBin.tsx

```diff
@@ -0,0 +1,411 @@
+          }),
+        };
+      }
+      return {
+        allValues: allColumns.numColVals[1].resolvedValues,
+        filteredValues: allColumns.numColVals[1].resolvedValues,
+      };
+    }
+
+    return null;
+  }, [allColumns, colsStatus, config.color, filteredCategories]);
+
+  // resize observer for setting size of the svg and updating on size change
+  useEffect(() => {
+    const ro = new ResizeObserver((entries: ResizeObserverEntry[]) => {
+      setHeight(entries[0].contentRect.height - margin.top - margin.bottom);
+      setWidth(entries[0].contentRect.width - margin.left - margin.right);
+    });
+
+    if (ref) {
+      ro.observe(ref.current);
+    }
+  }, []);
+
+  // create x scale
+  const xScale = useMemo(() => {
+    if (currentX?.allValues) {
+      const min = d3v7.min<number>(currentX.allValues.map((c) => c.val as number));
+      const max = d3v7.max<number>(currentX.allValues.map((c) => c.val as number));
+
+      const newScale = d3v7
+        .scaleLinear()
+        .domain([min, max])
+        .range([margin.left, margin.left + width]);
+
+      if (xRescaleFunc) {
+        xZoomedScale.current = xRescaleFunc(newScale);
+      }
+
+      return newScale;
+    }
+
+    return null;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [currentX?.allValues, width]);
+
+  // create y scale
+  const yScale = useMemo(() => {
+    if (currentY?.allValues) {
+      const min = d3v7.min<number>(currentY.allValues.map((c) => c.val as number));
+      const max = d3v7.max<number>(currentY.allValues.map((c) => c.val as number));
+
+      const newScale = d3v7
+        .scaleLinear()
+        .domain([min, max])
+        .range([margin.top + height, margin.top]);
+
+      if (yRescaleFunc) {
+        yZoomedScale.current = yRescaleFunc(newScale);
+      }
+
+      return newScale;
+    }
+
+    return null;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [currentY?.allValues, height]);
+
+  // creating d3 hexbin object to do hex math for me
+  const d3Hexbin = useMemo(() => {
+    return hex
+      .hexbin()
+      .radius(config.hexRadius)
+      .extent([
+        [0, 0],
+        [width, height],
+      ]);
+  }, [config.hexRadius, height, width]);
+
+  // generating the actual hexes
+  const hexes: HexbinBin<[number, number, string, string]>[] = useMemo(() => {
+    const inputForHexbin = [];
+
+    if (currentX && currentY) {
+      currentX.filteredValues.forEach((c, i) => {
+        inputForHexbin.push([
+          xScale(c.val as number),
+          yScale(currentY.filteredValues[i].val as number),
+          currentColorColumn ? currentColorColumn.filteredValues[i].val : '',
+          c.id,
+        ]);
+      });
+    }
+
+    // TODO: Im cheating a bit here by appending the id/color value to each hex, breaking the types.
+    // is there a better way to type this?
+    return d3Hexbin(inputForHexbin) as unknown as HexbinBin<[number, number, string, string]>[];
```

Comment:
As far as I understand, the original HexbinBin typing only supports `[number, number]` and you added color and id as additional information. Where are you using/reading the two additional properties? It doesn't seem to be consumed by the `.hexbin()` function above.

Reply:
I'm using them in the SingleHex component to assign categories/figure out if a hex is selected. D3 itself does not use those values at all, but it also does not remove them making them useful for me. Otherwise I'd need to keep a separate variable which is up to date with the d3 hexes that has this information.

Reply:
I see. I'm fine with this solution. The only thing I would change then is to remove the "TODO" + question. Then just document this addition in a comment and mention where it is currently used. This way future developer can easier follow-up and navigate in the code. Thanks.